### PR TITLE
Derive pkg from ansible facts

### PR DESCRIPTION
**Pull request by luckslovez**
Includes the commit in https://github.com/newrelic-experimental/caos-ansible-roles/pull/51

Currently the only way to derive the falcon-sensor pkg hash is using the hostname. This PR introduces an alternative method which uses ansible facts, in particular;
- `ansible_distribution`
- `ansible_major_version`
- `ansible_architecture`

Note that not all use cases have been covered, only those we (logging team) need, in particular;
- `ubuntu:16,18,20,22`
- `debian:10,11,12`
- `amazon-linux:2`
- `centos:7,8,9`
- `sles:12,15`
- `windows`

Implementation approach;
- Compute the falcon-sensor package hash early and populate it (this avoids looping)
- First try to match the hostname
- If no hostname matched then try with ansible facts
- Skip if nothing matched
- Select tasks file based on `ansible_os_family` and `ansible_pkg_mgr`

### CODE REVIEW

File: caos.ansible_roles/roles/install_crowdstrike_falcon/defaults/main.yaml [hunk 1]
```diff
@@ -0,0 +1 @@
+win_download_path: 'C:\Users\ansible\Downloads\falcon-sensor.exe'
```
Comment (luckslovez) [inline caos.ansible_roles/roles/install_crowdstrike_falcon/defaults/main.yaml:1]: Our usecase needs a different download path

File: caos.ansible_roles/roles/install_crowdstrike_falcon/tasks/zyp.yaml [hunk 2]
```diff
@@ -19,13 +24,11 @@
   zypper:
     name: libnl3-200
     state: present
-  when: inventory_hostname is search(item.hostname) and ansible_architecture == "aarch64"
-  loop: "{{ zyp }}"
+  when: ansible_architecture == "aarch64"
```
Comment (jsubirat) [inline caos.ansible_roles/roles/install_crowdstrike_falcon/tasks/zyp.yaml:27]: Nice one!

File: caos.ansible_roles/roles/install_crowdstrike_falcon/vars/main.yaml [hunk 3]
```diff
@@ -26,95 +26,215 @@ mac_hash: "cef4f59bc87e35fabe490e0f34b2f4dfaa076822e5a61d0a5fffc58014d446a1"
...
     hash: "{{ centos8_arm64_hash }}"
+    distribution: "CentOS"
+    major_version: "8"
+    architecture: "aarch64"
     # CentOS Stream not explicitly supported by CrowdStrike, using CentOS 8 installers instead
   - hostname: "amd64:centos-stream"
     hash: "{{ centos8_amd64_hash }}"
+    distribution: ""
+    major_version: ""
+    architecture: ""
```
Comment (jsubirat) [inline caos.ansible_roles/roles/install_crowdstrike_falcon/vars/main.yaml:42]: Instead of using the magic string `""` here, define a constant at the top of the file that makes it very clear why there is no value here:

```
# There are some distributions for which we haven't determined what is their `ansible_distribution`, 
# `ansible_distribution_major_version` and `ansible_architecture` facts. Once you have found out 
# these values for your AMIs, replace the values accordingly.
TODO_CHECK_ANSIBLE_FACTS_FOR_THIS_DISTRIBUTION: ""
```

And then use it here:

```
  - hostname: "amd64:centos-stream"
    hash: "{{ centos8_amd64_hash }}"
    distribution: "{{ TODO_CHECK_ANSIBLE_FACTS_FOR_THIS_DISTRIBUTION }}"
    major_version: "{{ TODO_CHECK_ANSIBLE_FACTS_FOR_THIS_DISTRIBUTION }}"
    architecture: "{{ TODO_CHECK_ANSIBLE_FACTS_FOR_THIS_DISTRIBUTION }}"
```

File: caos.ansible_roles/roles/install_crowdstrike_falcon/vars/main.yaml [hunk 4]
```diff
@@ -26,95 +26,215 @@ mac_hash: "cef4f59bc87e35fabe490e0f34b2f4dfaa076822e5a61d0a5fffc58014d446a1"
...
+    distribution: "Debian"
+    major_version: "12"
+    architecture: "aarch64"
 
 zyp:
   - hostname: "amd64:sles-12.3"
     hash: "{{ sles12_amd64_hash }}"
+    distribution: "SLES_SAP"
+    major_version: "12"
+    architecture: "x86_64"
```
Comment (jsubirat) [inline caos.ansible_roles/roles/install_crowdstrike_falcon/vars/main.yaml:186]: Hmmm unfortunately this is the one we're using for our tests (SLES distro with SAP installed). Probably CAOS could tell us what AMIs they are using, in case we can unify the images used for the testing, as I suspect these are quite Logging-specific. 🤔

File: caos.ansible_roles/roles/install_crowdstrike_falcon/defaults/main.yaml (hunk 1 above)
Comment (jsubirat) [reply caos.ansible_roles/roles/install_crowdstrike_falcon/defaults/main.yaml:1]: Why is that?

File: caos.ansible_roles/roles/install_crowdstrike_falcon/tasks/win.yaml [hunk 5]
```diff
@@ -11,29 +11,24 @@
       Accept: "application/json"
       Content-Type: "application/x-www-form-urlencoded"
   register: crowdstrike_auth
-  when: ansible_facts['os_family'] == 'Windows'
```
Comment (jsubirat) [inline caos.ansible_roles/roles/install_crowdstrike_falcon/tasks/win.yaml:14]: Revert this is if you revert the changes in `main.yml`.

File: caos.ansible_roles/roles/install_crowdstrike_falcon/tasks/win.yaml [hunk 6]
```diff
@@ -11,29 +11,24 @@
...
-    url: "{{ api_base_url }}/sensors/entities/download-installer/v1?id={{ item.hash }}"
+    url: "{{ api_base_url }}/sensors/entities/download-installer/v1?id={{ win[0].hash }}"
     url_timeout: 120
     headers:
       Authorization: "Bearer {{ crowdstrike_auth.json.access_token }}"
       Accept: "application/octet-stream"
       Content-Type: "application/json"
-    dest: C:\Users\ansible\Downloads\falcon-sensor.exe
-  when: ansible_facts['os_family'] == 'Windows' and inventory_hostname is search(item.hostname)
-  loop: "{{ win }}"
```
Comment (jsubirat) [inline caos.ansible_roles/roles/install_crowdstrike_falcon/tasks/win.yaml:26]: I'd probably revert this and still allow it to iterate through a list of supported packages

File: caos.ansible_roles/roles/install_crowdstrike_falcon/vars/main.yaml (hunk 4 above)
Comment (jsubirat) [reply caos.ansible_roles/roles/install_crowdstrike_falcon/vars/main.yaml:186]: Even if we use specific AMIs, this should not be a problem: we can just add them as separate items, so they are not tied to a particular hostname. In fact, thinking it better, I'd suggest to adopt this way instead:

```
IGNORE_HOSTNAME: "IGNORED"
IGNORE_ANSIBLE_FACTS: "IGNORED"

...
  zyp:
    - hostname: "amd64:sles-12.3"
      hash: "{{ sles12_amd64_hash }}"
      distribution: "{{ IGNORE_ANSIBLE_FACTS }}"
      major_version: "{{ IGNORE_ANSIBLE_FACTS }}"
      architecture: "{{ IGNORE_ANSIBLE_FACTS }}"
     - hostname: "{{ IGNORE_HOSTNAME }}"
      hash: "{{ sles12_amd64_hash }}"
      distribution: "SLES_SAP"
      major_version: "12"
      architecture: "x86_64"
```

Or even better, can we just allow two different schemas and allow empty values in the list objects?:
```
  zyp:
    - hostname: "amd64:sles-12.3"
      hash: "{{ sles12_amd64_hash }}"
     - distribution: "SLES_SAP"
      major_version: "12"
      architecture: "x86_64"
      hash: "{{ sles12_amd64_hash }}"
```

This way, if someone else needs to support another SLES12 distribution, they can just add it to the list.

**Review by jsubirat — COMMENTED** [review]
Some comments, @luckslovez

File: caos.ansible_roles/roles/install_crowdstrike_falcon/tasks/main.yaml [hunk 7]
```diff
@@ -14,11 +14,17 @@
   when: ansible_facts['os_family'] != 'Windows'
 
 - include_tasks: dnf.yaml
+  when: ansible_facts['os_family'] != 'Windows' and ansible_pkg_mgr == 'dnf'
```
[the remark this answers is not in the document]
Comment (luckslovez) [reply caos.ansible_roles/roles/install_crowdstrike_falcon/tasks/main.yaml:17]: That makes sense! Ansible facts don't come for granted 👍🏽

File: caos.ansible_roles/roles/install_crowdstrike_falcon/defaults/main.yaml (hunk 1 above)
Comment (luckslovez) [reply caos.ansible_roles/roles/install_crowdstrike_falcon/defaults/main.yaml:1]: It maybe has to do with the ami itself, but we log in as Admins and there is no other user apparently
image
`vars/main.yml` takes precedence over playbook variables, so I needed to create the CAOS' use case as default 🤔 
It might be a better way to achieve that, though 🙇🏽

File: caos.ansible_roles/roles/install_crowdstrike_falcon/vars/main.yaml (hunk 4 above)
Comment (luckslovez) [reply caos.ansible_roles/roles/install_crowdstrike_falcon/vars/main.yaml:190]: That is a good point since there are no raw sles15/12 specific SP distros available apparently. I agree that if CAOS is using a different distribution we should switch to it too.

**Review by DavSanchez — COMMENTED** [review]
Hi! First of all sorry for the delay. The changes overall look good to me and I have learned a bit more of Ansible in the process. However, when spinning-up a set of machines, I have found issues with the SLES ones that I'd like to troubleshoot deeplier before approving (they might be related to a recent change in the env provisioner: https://github.com/newrelic-experimental/env-provisioner/pull/18 and not to this PR). I'll try to review that and get back ASAP!

**Review by DavSanchez — APPROVED** [review]
I located the issue, added a quick PR and resolved conflicts here. I tested running canaries again and the provisioning went good.

LGTM! 🚀

**Comment by luckslovez** [comment]
@DavSanchez would it be possible for you to give this another test run? I had to pull main and want to make sure nothing went sideways. Otherwise this is good to merge 👍🏽 !

**Comment by DavSanchez** [comment]
Just ran it and looks like everything is fine. LGTM! 🚀 

Let me know if you are able to merge or if we have to do it from our end.

**Review by DavSanchez — APPROVED** [review]

**Review by DavSanchez — APPROVED** [review]

**Comment by DavSanchez** [comment]
Repolinter CI is stuck and cannot be forced to run apparently. Given it's the only CI that has not run and everything is fine otherwise, I will merge this. Thanks @luckslovez!
